dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
+	// returns a default value for each type:
+	// - empty string
+	// - zero
+	// - empty slice
+	// - 0.0.0 for a version
+	//
+	// `device.attributes` returns nil.
+	//
+	// These defaults are useful because CEL expression are applied to *all*
+	// available devices from any driver when the driver name is unset.
+	// Looking up an attribute defined by one vendor when checking a device
+	// from some other vendor doesn't find the attribute.
+	//
+	// Some examples:
+	//    "a.dra.example.com" in device.attributes &&
+	//    device.quantityAttributes["a.dra.example.com"].isGreaterThan(quantity("0")) &&
+	//    # No separate check, b.dra.example.com yields the empty slice when not set.
+	//    device.stringsliceAttributes["b.dra.example.com"].isSorted()
+	//
+	// Attribute names have to be fully-qualified.
+	//
+	// The `device.driverName` string variable can be used to check for a specific
+	// driver explicitly in a filter that is meant to work for devices from
+	// different vendors.
+	//
+	// If empty, any device matches.
+	//
+	// +optional
+	Selector string `json:"selector" protobuf:"bytes,2,name=selector"`
+}
+
+// FUTURE EXTENSION, not planned for 1.31! Needs further thought.
+type ResourceRequirement struct {
+	Name                     string `json:"name"`
+	ResourceRequirementValue `json:",inline" protobuf:"bytes,2,opt,name=value"`
+}
+
+// ResourceRequirementValue represents the value of a ResourceRequirement.
+// ResourceRequirementValue must have one and only one field set.
+type ResourceRequirementValue struct {
+	// QuantityValue is a quantity.
+	QuantityValue *resource.Quantity `json:"quantity,omitempty" protobuf:"bytes,1,opt,name=quantity"`
+
+	// IntRangeValue is a range of 64-bit integers.
+	IntRangeValue *IntRange `json:"intRange,omitempty" protobuf:"varint,2,rep,name=intRange"`
 }
 
 // Namespace scoped.
-type DeviceClaim struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   DeviceClaimSpec   `json:"spec,omitempty"`
-	Status DeviceClaimStatus `json:"status,omitempty"`
+// ResourceClaim describes which resources (typically one or more devices)
+// are needed by a resource consumer.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClaim struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	ResourceClaimSpecAlternatives `json:",inline"` // Inlined with "spec" defined in the ResourceClaimSpec.
+
+	// Status describes whether the claim is ready for use.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ResourceClaimSpecAlternatives defines how a claim is to be allocated.
+type ResourceClaimSpecAlternatives struct {
```

Comment:
I am having trouble following all of this...

Reply:
In a nutshell, this allows writing "normal" claim objects:
```yaml
kind: ResourceClaim
...
spec:
  requests:
  - ...
```

The "spec" in this "normal" case is the `ResourceClaimSpecAlternatives.Spec` pointer.

The alternative is to use a vendor CRD:
```yaml
kind:  ResourceClaim
...
specRef:
  name: my-claim-spec
  kind: ClaimSpec
  apiGroup: dra.example.com
```

The mechanism here is the same as it was for claim and class parameters in 1.30, except that now it gets applied to an entire claim (and only there).

Reply:
Focusing on "this is the spec" is absolutely the right way to approach this.

Reply:
In general, I am OK with not including this for now. It will only become _essential_ if / when we allow custom controllers to make allocation decisions using structured parameters (i.e. in conjunction with the in-tree scheduler -- not via "classic" DRA). Such controllers will want to express something beyond what is possible with the normal set of fields in a Claim (why else would they need to be custom), so will need a way of expressing this API via a custom CRD.

Note that this also has implications for continued support of "classic" DRA in general.

Without the ability to provide a custom CRD (either as a full ClaimSpec or as ClaimParameters (which we don't have anymore)), users of "classic" DRA will have to abuse the `ConfigurationParameters.Vendor` field to embed their CRDs. I say "abuse" because this field is only meant for providing configuration information to the kubelet plugin of a DRA driver, but users of "classic" DRA will also need to include selection information in it.

I think this is probably "fine" since DRA as a whole is still in alpha and anyone currently using it needs to adapt to whatever changes we make. However, if we have a long-term plan for continued support of "classic" DRA, then this should probably be added sooner rather than later.

Reply:
Based on @klueska's comment, I am taking it out. This also removes the `ResourceClaimSpecification` type.

We can describe it as a field that is feature-gated by classic DRA, without allowing its use with the Kubernetes scheduler. The custom controllers can still have their own CRD-based API.